transports/janus_mqtt.c

```diff
@@ -0,0 +1,648 @@
+ * \author Andrei Nesterov <ae.nesterov@gmail.com>
+ * \copyright GNU General Public License v3
+ * \brief  Janus MQTT transport plugin
+ * \details  This is an implementation of a MQTT transport for the Janus API,
+ * using the Eclipse Paho C Client library (https://eclipse.org/paho/clients/c).
+ * This means that this module adds support for MQTT based messaging as
+ * an alternative "transport" for API requests, responses and notifications.
+ * This is only useful when you're handling the communication with
+ * clients your own way. Right now, you can only configure
+ * the address of the MQTT broker to use, and the queues to
+ * make use of to receive (to-janus) and send (from-janus) messages
+ * from/to an external application. As with WebSockets, considering that
+ * requests wouldn't include a path to address some mandatory information,
+ * these requests addressed to Janus should include as part of their payload,
+ * when needed, additional pieces of information like \c session_id and
+ * \c handle_id. That is, where you'd send a Janus request related to a
+ * specific session to the \c /janus/<session> path, with RabbitMQ
+ * you'd have to send the same request with an additional \c session_id
+ * field in the JSON payload.
+ * \note When you create a session using RabbitMQ, a subscription to the
+ * events related to it is done automatically through the outgoing queue,
+ * so no need for an explicit request as the GET in the plain HTTP API.
+ *
+ * \ingroup transports
+ * \ref transports
+ */
+
+#include "transport.h"
+#include "../debug.h"
+#include "../config.h"
+#include "../utils.h"
+#include <MQTTAsync.h>
+
+/* Transport plugin information */
+#define JANUS_MQTT_VERSION        1
+#define JANUS_MQTT_VERSION_STRING "0.0.1"
+#define JANUS_MQTT_DESCRIPTION    "This transport plugin adds MQTT support to the Janus API via Paho client library."
+#define JANUS_MQTT_NAME           "JANUS MQTT transport plugin"
+#define JANUS_MQTT_AUTHOR         "Andrei Nesterov <ae.nesterov@gmail.com>"
+#define JANUS_MQTT_PACKAGE        "janus.transport.mqtt"
+
+/* Transport methods */
+janus_transport *create(void);
+int janus_mqtt_init(janus_transport_callbacks *callback, const char *config_path);
+void janus_mqtt_destroy(void);
+int janus_mqtt_get_api_compatibility(void);
+int janus_mqtt_get_version(void);
+const char *janus_mqtt_get_version_string(void);
+const char *janus_mqtt_get_description(void);
+const char *janus_mqtt_get_name(void);
+const char *janus_mqtt_get_author(void);
+const char *janus_mqtt_get_package(void);
+gboolean janus_mqtt_is_janus_api_enabled(void);
+gboolean janus_mqtt_is_admin_api_enabled(void);
+int janus_mqtt_send_message(void *context, void *request_id, gboolean admin, json_t *message);
+void janus_mqtt_session_created(void *context, guint64 session_id);
+void janus_mqtt_session_over(void *context, guint64 session_id, gboolean timeout);
+
+/* Transport setup */
+static janus_transport janus_mqtt_transport_ =
```

Comment:
I can't say I'm really a fan of variable names that end with an underscore, but that may be just me... :wink: